### PR TITLE
[SPARK-34263][SQL] Simplify the code for treating unicode/octal/escaped characters in string literals

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

In the current master, the code for treating unicode/octal/escaped characters in string literals is a little bit complex so let's simplify it.

### Why are the changes needed?

To keep it easy to maintain.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

`ParserUtilsSuite` passes.

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
Looks fine if logic is the same and tests pass. It might also be faster.

**Comment by srowen** [comment]
You're welcome to merge when ready.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 1]
```diff
@@ -187,33 +178,19 @@ object ParserUtils {
...
-          i += 1
-        }
-      } else {
-        // non-escaped character literals.
-        sb.append(currentChar)
+          appendEscapedChar(c.charAt(0))
+          charBuffer.position(charBuffer.position() + 2)
+        case _ =>
+          // non-escaped character literals.
+          sb.append(charBuffer.get())
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:192]: Just out of curiosity; the performance of the `non-escpaed` long string case is almost the same before/after this PR? This improvement itself looks fine though.

**Comment by sarutak** [comment]
@srowen I noticed an incompatible behavior and fixed it (also added an assertion for that case).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala (hunk 1 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:192]: I think it's almost the same. Just checked with `SELECT "(10000 characters)";` and I recognized no remarkable difference.

**Review by srowen — APPROVED** [review]

**Comment by sarutak** [comment]
Merged to `master`. Thanks all!

**Pull request closed without merging**